### PR TITLE
C#: Add flow steps from a PageModel to cshtml page.

### PR DESCRIPTION
**Pull request by joefarebrother**
Adds flow steps from `PageModel` handler methods (calls to [`PageModel.Page()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razorpages.pagemodel.page?view=aspnetcore-7.0#microsoft-aspnetcore-mvc-razorpages-pagemodel-page) and [`PageModel.RedirectToPage()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razorpages.pagemodel.redirecttopage?view=aspnetcore-7.0#microsoft-aspnetcore-mvc-razorpages-pagemodel-redirecttopage) as well as implicit flow from `this` in void-returning handler methods) to the `Model` field in the code generated from the `.cshtml` file for the corresponding page.

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (tamasvajk) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:228]: Should `OnPageHandler%` methods be filtered out?

**Review by tamasvajk — APPROVED** [review]
I don't know enough about razor pages, but this looks plausible to me. I added one minor comment. 

Let's wait for others to review it before merging.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:228]: Hmm, I guess those methods are not technically handler methods; however they are called during the page handler process, so it makes sense to consider tainted writes to the model in those methods to still propagate to the page.

File: csharp/ql/test/query-tests/Security Features/CWE-079/XssPageModels/test.ql
Comment (hvitved) [inline csharp/ql/test/query-tests/Security Features/CWE-079/XssPageModels/test.ql:19]: I suggest making this a `path-problem` query instead.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:257]: Why only implicit `this` qualifiers?

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll:264]: For the ` getImplicitThisCallInVoidHandler` case, I would expect that the node should be the post-update node, as that is what is that target of stores. I.e., in `this.Name = source()` the store is into the post-update node for `this`.

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll]: Updated to count all `this` qualifiers

File: csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll
Comment (joefarebrother) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll]: How can I target specifically the post-update node?

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll]: You need to import `dataflow.internal.DataFlowPrivate.qll` and then do something like
```ql
private ThisAccess getThisCallInVoidHandler(PageModelClass pm) {
  exists(MethodCall mc |
    mc.getEnclosingCallable() = pm.getAHandlerMethod() and
    mc.getEnclosingCallable().getReturnType() instanceof VoidType and
    result = mc.getQualifier()
  )
}

private class PageModelJumpNode extends DataFlow::NonLocalJumpNode {
  PageModelClass pm;

  PageModelJumpNode() {
    this.asExpr() = getAPageCall(pm).getQualifier()
    or
    this.(PostUpdateNode).getPreUpdateNode().asExpr() = getThisCallInVoidHandler(pm)
  }
```

In case you are wondering, the test example works (by accident) because
```csharp
    public void OnGet() { 
        Name = source();
    }
```

is really

```csharp
    public void OnGet() { 
        this.Name = this.source();
    }
```

and there is use-use flow from the post-update `this` node in `this.Name` to the normal `this` node in `this.source()`, and then the current logic works.

**Review by tamasvajk — APPROVED** [review]
LGTM

**Pull request merged**
